src/server/worker/pipeline/transform/registry.go

```diff
@@ -0,0 +1,951 @@
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := r.Close(); err != nil && retErr != nil {
+			retErr = err
+		}
+	}()
+	pbr := pbutil.NewReader(r)
+	datums := make(map[string]struct{})
+	for {
+		k, err := pbr.ReadBytes()
+		if err != nil {
+			if err == io.EOF {
+				return datums, retErr
+			}
+			return nil, err
+		}
+		datums[string(k)] = struct{}{}
+	}
+}
+
+// Walk from the given commit back to a successfully completed commit so we can
+// get the initial state of datumsBase in the registry.
+func (reg *registry) getParentCommitInfo(commitInfo *pfs.CommitInfo) (*pfs.CommitInfo, error) {
+	pachClient := reg.driver.PachClient()
+	outputCommitID := commitInfo.Commit.ID
+
+	// Walk up the commit chain to find a successfully finished commit
+	for commitInfo.ParentCommit != nil {
+		reg.logger.Logf(
+			"blocking on parent commit %q before writing to output commit %q",
+			commitInfo.ParentCommit.ID, outputCommitID,
+		)
+		parentCommitInfo, err := pachClient.PfsAPIClient.InspectCommit(pachClient.Ctx(),
+			&pfs.InspectCommitRequest{
+				Commit:     commitInfo.ParentCommit,
+				BlockState: pfs.CommitState_FINISHED,
+			})
+		if err != nil {
+			return nil, err
+		}
+		if parentCommitInfo.Trees != nil {
+			return parentCommitInfo, nil
+		}
+		commitInfo = parentCommitInfo
+	}
+	return nil, nil
+}
+
+// ensureJob loads an existing job for the given commit in the pipeline, or
+// creates it if there is none. If more than one such job exists, an error will
+// be generated.
+func (reg *registry) ensureJob(
+	commitInfo *pfs.CommitInfo,
+	metaCommit *pfs.Commit,
+) (*pps.JobInfo, error) {
+	pachClient := reg.driver.PachClient()
+
+	// Check if a job was previously created for this commit. If not, make one
+	jobInfos, err := pachClient.ListJob("", nil, commitInfo.Commit, -1, true)
+	if err != nil {
+		return nil, err
+	}
+	if len(jobInfos) > 1 {
+		return nil, fmt.Errorf("multiple jobs found for commit: %s/%s", commitInfo.Commit.Repo.Name, commitInfo.Commit.ID)
+	} else if len(jobInfos) < 1 {
+		job, err := pachClient.CreateJob(reg.driver.PipelineInfo().Pipeline.Name, commitInfo.Commit, metaCommit)
+		if err != nil {
+			return nil, err
+		}
+		reg.logger.Logf("created new job %q for output commit %q", job.ID, commitInfo.Commit.ID)
+		// get jobInfo to look up spec commit, pipeline version, etc (if this
+		// worker is stale and about to be killed, the new job may have a newer
+		// pipeline version than the master. Or if the commit is stale, it may
+		// have an older pipeline version than the master)
+		return pachClient.InspectJob(job.ID, false)
+	}
+
+	// get latest job state
+	reg.logger.Logf("found existing job %q for output commit %q", jobInfos[0].Job.ID, commitInfo.Commit.ID)
+	return pachClient.InspectJob(jobInfos[0].Job.ID, false)
+}
+
+func (reg *registry) startJob(commitInfo *pfs.CommitInfo, metaCommit *pfs.Commit) error {
+	pachClient := reg.driver.PachClient()
+
+	if reg.datumsBase == nil {
```

Comment:
Something I like to do with this kind of stuff is to actually have the check in the function rather than the calling function. So `initializeDatumsBase` would check if `datumsBase` is nil. I think this is a bit cleaner.